cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Utility methods for decomposing arbitrary n-qubit (2^n x 2^n) unitary."""
+
+from scipy.linalg import cossin
+
+import numpy as np
+
+import cirq
+
+
+def quantum_shannon_decomposition(qubits: list, U: np.ndarray, ops=None):
+    """Returns a list of operations for an arbitrary n-qubit decomposition, preserving phase
+
+    The algorithm is described in Shende et al.:
+    Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+    https://arxiv.org/abs/quant-ph/0406176
+
+    Args:
+        qubits: List of qubits in order of significance
+        U: Numpy array for unitary matrix representing gate to be decomposed
+        ops: List of new existing operations on which to append new operations, whenever
+                a recusive call is made
+             If 'None' is given, a new list is instantiated
```

Comment:
Instead of accepting `ops` as a mutable argument to the method and mutating it as part of the implementation; always return the OP_TREE corresponding to this decomposition and append it to the list of `ops` at the callsite, if needed. 

This means changing the function signature to 

```python
def quantum_shannon_decomposition(qubits: list, u: np.ndarray) -> List[cirq.Operation]:
```

Also do this change for all the helper methods, like `_single_qubit_decomposition`, `_msb_demuxer` and `_multiplexed_cossin`

Reply:
Agreed- I like this approach. I added a return for top-level function. Converter rest to generators to yield output